climada_petals/hazard/copernicus_forecast/handler.py

```diff
@@ -0,0 +1,560 @@
+"""
+This file is part of CLIMADA.
+
+Copyright (C) 2017 ETH Zurich, CLIMADA contributors listed in AUTHORS.
+
+CLIMADA is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free
+Software Foundation, version 3.
+
+CLIMADA is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along
+with CLIMADA. If not, see <https://www.gnu.org/licenses/>.
+
+---
+
+Module to handle seasonal forecast data from the Copernicus Climate Data Store (CDS)
+in the U-CLIMADAPT project.
+
+TODO: update prerequisites
+Prerequisites:
+1. CDS API client installation:
+   pip install cdsapi
+
+2. CDS account and API key:
+   Register at https://cds-beta.climate.copernicus.eu
+
+3. CDS API configuration:
+   Create a .cdsapirc file in your home directory with your API key and URL.
+   For instructions, visit:
+   https://cds-beta.climate.copernicus.eu/how-to-api#install-the-cds-api-client
+"""
+
+import os
+import logging
+import calendar
+import re
+
+import xarray as xr
+import pandas as pd
+import numpy as np
+import cdsapi
+
+from climada.hazard import Hazard
+from climada.util.coordinates import get_country_geometries
+import climada_petals.hazard.copernicus_forecast.indicator as indicator
+
+
+LOGGER = logging.getLogger(__name__)
+
+class ForecastHandler:
+    """
+    A class to handle downloading, processing, and calculating climate indices
+    and hazards based on seasonal forecast data from Copernicus Climate Data Store (CDS).
+    """
+
+    _FORMAT_GRIB = 'grib'
+    _FORMAT_NC = 'nc'
+    
+    def __init__(self, data_dir='.', url = None, key = None):
+        """
+        Initializes the ForecastHandler instance.
+
+        Parameters:
+        data_dir (str): Path to the directory where downloaded and processed data will be stored.
+            Defaults to the current directory ('.').
+        url (str): url to the CDS API. Defaults to None, in which case the url from /.cdsapirc
+            is used.
```

Comment:
Why is there the need to give in the url? Is there more than one possible?

Reply:
I have maintained the current structure for the Copernicus Data Store (CDS), which includes the option to use a configurable URL. However, CDS Beta, the new Copernicus website, has recently been launched last week, and many components are undergoing changes. The idea is to keep this setup until everything stabilizes and we have a clearer understanding of how to optimize downloads under the new circumstances since more things can change.

Reply:
OK, then I would put the default as the string instead of None. Also, it seems that this class attribute is never used later on.

Reply:
We put it to None because when handing "None" to the client.retreive function of the cdsapi module, the client will use the url and key that are saved in a separate file in the user's home directory (see instructions). If I understood correctly, this file is necessary for the downloader to work, so the user must add it anyways.

Reply:
Ok, interesting. Then the docstring is incorrect. Overall though, I think this could be handled more smoothly.

Reply:
Also to add, Dahyann said that DWD had specific requests on what the user should input when they want to download something (they wanted that the users have the cdsapi config file and ALSO have to put in url and key by hand). Then, as Dahyann said, the API is currently changing and we don't know the final form. 
So we agreed that we leave it as is for now and when we know all boundary conditions, we could optimize further. I agree with you that it could be handled better

Reply:
I understand that, and it is good to listen to DWD. At the same time in the end, this is a CLIMADA module, not a DWD module. Thus, we should still keep a sensible coding setting whenever possible.

Reply:
Here it is a question of design, and not really of optimization. Thus I would caution against using the argument: "we will just fix it later".

Reply:
ok agreed, we will have a chat about better possibilities.

Reply:
I removed the URL parameter to avoid confusion. The default .cdsapirc configuration is sufficient for standard use since there's "usually" only one .cdsapirc URL.